core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -279,27 +262,31 @@ class TransactionCoordinator(brokerId: Int,
   private def logInvalidStateTransitionAndReturnError(transactionalId: String,
                                                       transactionState: TransactionState,
                                                       transactionResult: TransactionResult) = {
-    error(s"transactionalId: $transactionalId -- Current state is $transactionState, but received transaction " +
-      s"marker result: $transactionResult")
+    debug(s"TransactionalId: $transactionalId's state is $transactionState, but received transaction " +
+      s"marker result to send: $transactionResult")
     Left(Errors.INVALID_TXN_STATE)
   }
 
   def handleEndTransaction(transactionalId: String,
                            producerId: Long,
                            producerEpoch: Short,
                            txnMarkerResult: TransactionResult,
-                           responseCallback: EndTxnCallback): Unit = {
-    val error = validateTransactionalId(transactionalId)
-    if (error != Errors.NONE)
-      responseCallback(error)
+                           responseCallback: EndTxnCallback,
+                           checkTxnTimeout: Boolean = false): Unit = {
```

Comment:
This is ok, but it seems it's simpler to check the expiration time in one place in TransactionStateManager.transactionsToExpire().